### DISCUSSION TITLE
Does MsQuic have support for macOS?

### DISCUSSION THREAD

**Discussion by KTSnowy** [microsoft/msquic#3594] [Q&A] [answered]
We're attempting to define a new protocol specialized for tag-based message passing over QUIC (without the extra overhead of a full HTTP/3 server). This will be used in [our COBOL compiler](https://github.com/otterkit/otterkit) to provide support for the new async messaging features added in the COBOL 2023 standard.

But we're not sure if MsQuic supports running both servers and clients on macOS. The [Platform.md](https://github.com/microsoft/msquic/blob/main/docs/Platforms.md) file doesn't mention macOS support, but the [latest release](https://github.com/microsoft/msquic/releases/tag/v2.1.8) does have a macOS universal build which would suggest that it does support it?

Because our compiler and runtime library are built almost entirely with C# we would like to try using System.Net.Quic, but the [documentation page](https://learn.microsoft.com/en-us/dotnet/core/extensions/httpclient-http3#platform-dependencies) for it also mentions macOS not being supported yet.

We looked at other implementations like Quiche from Cloudflare since it does seem to have support for macOS, but the time and complexity required to maintain C# bindings for it (specially because of Rust) would not be ideal for our small team.

Is there support for both server and clients on macOS? If not are there plans to officially support it in both MsQuic and System.Net.Quic?

**Comment by nibanks** [comment]
Generally, yes it works. Our main problems with macOS support have been:

1. No real official requirement from anyone.
2. We don't have a lot of expertise with macOS (nor really any physical hw to repro issues locally).

If you _could help_ with both above, it would make it easier to officially support macOS.

**Comment by KTSnowy** [comment]
> If you could help with both above, it would make it easier to officially support macOS.

Would Otterkit having that requirement count as a real official requirement from someone?

> We don't have a lot of expertise with macOS (nor really any physical hw to repro issues locally).

I have a MacBook Air and can help with local testing and setting up automated GitHub actions testing with macOS if needed.

Let me know in what way I could help and what needs to be done, I'll be happy to help.

**Reply by nibanks** [reply]
> Would Otterkit having that requirement count as a real official requirement from someone?

It'd definitely help. I guess I still don't quite understand how hard you dependency will be, or what your scenario entails.

> Let me know in what way I could help and what needs to be done, I'll be happy to help.

This helps as well. Perhaps one way to look at this: If you encounter a macOS specific problem with MsQuic, are you willing to help debug and possibly fix it?

**Reply by KTSnowy** [reply]
> It'd definitely help. I guess I still don't quite understand how hard you dependency will be, or what your scenario entails.

We're working on defining the spec for the new specialized message passing protocol and have made a [design doc draft](https://docs.google.com/document/d/1aqOZOHa8_TGMmBYEkQ0UvGhajWA4UME17Zp7sJKQsoA/edit#) which details most of the technical requirements for our scenario (and for COBOL standard conformance). 

This spec is very early and there are still things missing and a ton of issues with it, for example the version request should probably be moved to the application-layer protocol negotiation instead of being a separate request, and the authentication request could (maybe?) also be moved into the message tags instead of being an additional request.

We'll be working on improving the spec for this, but need a QUIC implementation to experiment with during development, and also one that can be linked into our runtime library.

> If you encounter a macOS specific problem with MsQuic, are you willing to help debug and possibly fix it?

Yes, I'll do my best. We plan for our compiler to work on Windows, Linux and macOS so we'll have to runs tests on all three. I can help debug and fix macOS specific issues if I encounter any during development.

Would you recommend us using the macOS release build [from here](https://github.com/microsoft/msquic/releases/tag/v2.1.8)?

**Reply by nibanks** [reply, marked as answer]
> Would you recommend us using the macOS release build [from here](https://github.com/microsoft/msquic/releases/tag/v2.1.8)?

Yes, it should be fine to use.

**Reply by KTSnowy** [reply]
Oh also, when using the Linux release from that page, are we required to also install the libmsquic package separately?

**Reply by nibanks** [reply]
You don't _have_ to use the package. You can include the library with your product as well.
